### ISSUE TITLE
Deadlock between ICache.Clear() and TryGet(IContext context)

### ISSUE THREAD

**Issue by bungeemonkee** [ninject/Ninject#182]
I know there are a couple other bugs that have been opened or closed about deadlocks and many were closed as simply being bad usage but I'm pretty sure that A) this is unique and B) this is not a usage issue.

I think I've found a deadlock that only occurs when calling Clear() on the cache for an IKernel in one thread while another is calling either Get<>() or GetAll<>().

Basically I have a custom scope provider that is essentially thread scoped except that it can also "increment" ie take the current thread scope, close it, and create a new scope for that thread. This allows me do to things like have a single thread perform multiple units of work in a loop with each unit having it's own scope.

Inside that scope provider, when tearing down the previous scope there is a call to clear the scope cache like this:

``` c-sharp
kernel.Components.Get<ICache>().Clear(scope);
```

Where kernel is my IKernel (a static reference - one per process) and scope is the object representing the current scope for the current thread.

When more than one thread is running at a time I inevitably get deadlocks. Most threads are waiting somewhere on a Get<>() or GetAll<>() but there is always at least one (usually just one) waiting on that call to Clear().

Unfortunately I can't seem to create a simple test case to produce the deadlock and the application where it happens is quite large and complex and interacts with a fairly complex db setup.

I only noticed it recently when we did a major overhaul that improved performance by about 2 orders of magnitude so there were WAY more threads calling into Ninject at the same time (rather than working on all our crummy BL).

I was able to work around it by putting a simple lock on the IKernel reference around the call to Clear() like this:

``` c-sharp
lock (kernel)
{
    kernel.Components.Get<ICache>().Clear(scope);
}
```

But that doesn't feel like a real solution to me.

**Comment by schambers** [comment]
I actually just came across this same problem but with [OnePerHttpRequestModule](https://github.com/ninject/Ninject.Web.Common/blob/master/src/Ninject.Web.Common/OnePerRequestHttpModule.cs) 

Specifically when in the anonymous method passed to MapKernels inside deactivation of the http request.

The only way I could reliably reproduce is purposefully throw exceptions in rapid succession amongst multiple threads until I could get a deadlock.

Is anyone aware of this issue? It's quite deep

**Comment by drieseng** [comment]
This is fixed in the upcoming 4.0 beta by https://github.com/ninject/Ninject/commit/7f4bcd0d7bd267acf7d959cf4d1b8ebe25876c06 and https://github.com/ninject/Ninject/commit/ca9a4c20ae6c25c7b1d67e6852516a9ecb054ab7.
